lms/models/h_group.py

```diff
@@ -9,15 +11,54 @@ class HGroup(NamedTuple):
     def groupid(self, authority):
         return f"group:{self.authority_provided_id}@{authority}"
 
+    @classmethod
+    def from_lti_parts(
+        # pylint: disable=too-many-arguments
+        cls,
+        name,
+        tool_consumer_instance_guid,
+        context_id,
+        section_id=None,
+        type_=type,
+    ):
+        """
+        Create an HGroup from LMS specific parts.
+
+        :param name: The name of the course
+        :param tool_consumer_instance_guid: Tool consumer GUID
+        :param context_id: Course id
+        :param section_id: A section id for a section group
+        :param type_: Group type (defaults to "course_group")
+        """
+        return HGroup(
+            name=cls._name(name) if name is not None else None,
+            authority_provided_id=cls._authority_provided_id(
+                tool_consumer_instance_guid, context_id, section_id
+            ),
+            type=type_,
+        )
```

Comment:
I don't think "lti parts" is quite the right term:

* Name isn't always an LTI thing (it does come from an LTI param for course groups, for section groups it comes from a non-LTI Canvas API param)
* `section_id` is a Canvas API thing not an LTI thing
* `type` is unrelated to LTI

But I'm not sure what a good name would be. Something like "group params" isn't very good.

I think PyLint's too-many-arguments warning may suggest a missing class:

```python
class HGroupParams(NamedTuple):
    """Params from which an `HGroup` can be created."""
    name: str
    tool_consumer_instance_guid: str
    context_id: str
    section_id: str=None
    type: str = "course_group"  # `type` default can now be removed from HGroup
```

But adding a class just to hold the params of a method, seems wrong.

So to be honest, at the cost of just a little duplication (almost none really), I think you can neatly avoid both issues and have nice meaningful `course_group()` and `section_group()` method names:

```python
    @classmethod
    def course_group(cls, name, tool_consumer_instance_guid, context_id):
        return HGroup(
            cls._name(name),
            hashed_id(tool_consumer_instance_guid, context_id),
        )

    @classmethod
    def section_group(cls, name, tool_consumer_instance_guid, context_id, section_id):
        return HGroup(
            cls._name(name),
            hashed_id(tool_consumer_instance_guid, context_id, section_id),
            type="section_group",
        )

    @classmethod
    def _name(cls, name):
        """Return an h-compatible group name from the given string."""
        if name is None:
            return None

        ...
```

(And note that `HGroup._authority_provided_id()` is now no longer needed.)


This would look pretty nice in the code that calls it:

```python
return HGroup.course_group(
    name=params["context_title"],
    tool_consumer_instance_guid=params["tool_consumer_instance_guid"],
    context_id=params["context_id"],
)

return HGroup.section_group(
    name=section.get("name"),
    tool_consumer_instance_guid=tool_consumer_instance_guid,
    context_id=context_id,
    section_id=section["id"],
)
```

Reply:
Yeah, that's better

Reply:
This is split up now and has a few benefits:

 * Type management is now an internal affair. You can't create a course group with a section type, or a random string
 * We can now name "name" as "course_name" / "section_name" which is more self documenting
 * It's also more accurate, as it's _not_ the eventual group name that is derived from it in all cases
 * Makes reading the places that create it much more obvious in intent as well